lib/absinthe/schema.ex

```diff
@@ -533,14 +533,10 @@ defmodule Absinthe.Schema do
   """
   @spec used_types(t) :: [Type.t()]
   def used_types(schema) do
-    [:query, :mutation, :subscription]
-    |> Enum.map(&lookup_type(schema, &1))
-    |> Enum.concat(directives(schema))
-    |> Enum.filter(&(!is_nil(&1)))
-    |> Enum.flat_map(&Type.referenced_types(&1, schema))
-    |> MapSet.new()
+    schema.__absinthe_types__
+    |> Map.keys()
     |> Enum.map(&Schema.lookup_type(schema, &1))
-    |> Enum.filter(&(!Type.introspection?(&1)))
+    |> Enum.filter(&(!Type.introspection?(&1) && &1.__private__[:__absinthe_used__]))
```

Comment:
We still get some nice performance benefits here because this just becomes a simple boolean filter instead of the complicated tree walk.

I do wonder if we should deprecate this function, I'm not sure that it has any use outside of trying to show stuff for introspection. Particularly if we can eliminate unused types from the compiled schema altogether I just don't see any point for it.

Reply:
I know we call it in our codebase since we tweak Introspection to further filter results, so even if it's just for modifying Introspection, it's still useful...

Reply:
Sorry to be clear I agree that making it easy to get introspectable types is useful, I just wonder if the function would be better named `introspectable_types` or something of that sort. `used_types` is sorta misleading since there wouldn't be any _unused_ types anymore, and introspection types are definitely used, they just aren't introspectable.

Reply:
Gotcha, that makes sense. The latest spec has this to say:

> When returning the set of types from the __Schema introspection type, all referenced built‐in scalars must be included. If a built‐in scalar type is not referenced anywhere in a schema (there is no field, argument, or input field of that type) then it must not be included.
> 
> When representing a GraphQL schema using the type system definition language, all built‐in scalars must be omitted for brevity.

https://spec.graphql.org/draft/#sec-Scalars.Built-in-Scalars

Given the spec language, maybe it's `referenced_types` that we want, and the flag in the structs should be `referenced`.

And since it's mentioned in the spec, do we want to have `sdl_types` that will just do the right thing? ("referenced_types - builtin")

Do you want to remove `used_types` or soft-deprecate?

Reply:
Gonna use referenced_types as the term. Instead of flagging the types, I'm currently doing:
```
        def __absinthe_types__() do
          __absinthe_types__(:used)
        end

        def __absinthe_types__(:used) do
          unquote(Macro.escape(used_types))
        end

        def __absinthe_types__(:all) do
          unquote(Macro.escape(type_list))
        end
```

This way `__absinthe_types__` by default just refers to referenced types.